### PR TITLE
fix : implement_try_eval_mode_arithmetic

### LINKED ISSUE #2021: try_ arithmetic functions return incorrect results
### Describe the bug

As part of exploring writing unit tests for serde code in https://github.com/apache/datafusion-comet/issues/2020, I discovered that we currently have incorrect behavior for `try_add` (and probably for many other `try_` functions).

The issue can be reproduced with the following test:

```scala
  test("try_add") {
    val data = Seq((Integer.MAX_VALUE, 1))
    withParquetTable(data, "tbl") {
      checkSparkAnswerAndOperator("SELECT try_add(_1, _2) FROM tbl")
    }
  }
```

This fails with:

```
== Results ==
!== Correct Answer - 1 ==      == Spark Answer - 1 ==
 struct<try_add(_1, _2):int>   struct<try_add(_1, _2):int>
![null]                        [-2147483648]
```

The issue is that we do not respect the `EvalMode` for this (and other) expressions. When serializing the expression we populate a `fail_on_error` flag based on `add.evalMode == EvalMode.ANSI`. However, we should really be serializing the `evalMode` and implementing all three distinct behaviors.

In the short term, we should probably just fall back to Spark for `EvalMode.TRY`, since the behavior is incorrect.

### Steps to reproduce

_No response_

### Expected behavior

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by coderfender**
## Which issue does this PR close?

Closes #2021

## Rationale for this change

 PR to support Try eval mode in native . Unfortunately neither DataFusion not Arrow crates support null on overflow ops which is the desired outcome in Spark (when Eval mode is set to Try)

## What changes are included in this PR?

1. New UDF called `checked_arithmetic` which would perform checked_add / checked_subtract/ checked_multiply over operands and wraps overflow in a NULL . (There aren't direct DataFusion options / Arrow Kernel APIs which would provide this functionality out of the box and hence the need for custom kernel + UDF based solution)
2. On the Spark side, the check to fallback to Spark when the EVAL Mode is set to Try is removed for above arithmetic ops.

## How are these changes tested?
1. Implemented unit tests with various overflow edge cases (add  , subtract , multiple , divide etc)

### CODE REVIEW

**Comment by coderfender** [comment]
Hello @andygrove , I Implemented custom arrow kernels to perform checked_add , checked_sub and checked_mul (registered as UDFS) supporting Integral types only (similar to spark's behavior) . My hope is to repeat / reuse this for other ops (in future), now that there is a framework established.

Please take a look whenever you get a chance and I can make changes (if any) to support try eval mode. Thank you very much

**Comment by coderfender** [comment]
@andygrove  , 
Here is the summary of changes : 
1. Spark's Try eval mode returns NULL in case there is a computation failure. Note that this is only supported / useful when the operands are integer (Int / Long) type since overflow on Float , Double and Decimal are non deterministic and/or return Nan/ Inf.
2. Since neither data fusion nor Arrow kernels have native implementation of Spark's `Try` eval mode , I went ahead and implemented custom UDFs (with custom Arrow kernels)  which perform checked_add, checked_sub, checked_mul , checked_div which return None when overflow occurs. 
3. I also verified that `div` and `integer_div` do work with the Try mode and added tests to check all possible edge cases.
4. `fail_on_overflow` param is added to `create_physical_expr` function to fork code to call UDF based on the Eval Option selected. 
5. Once this PR gets approved / merged, I will go ahead and continue to use this framework to implement the EVAL mode for other ops such as cast while also refactoring `abs` , `modulo` operations

**Comment by andygrove** [comment]
@coderfender looks like there are some clippy issues to be resolved

File: native/core/src/execution/planner.rs [hunk 1]
```diff
@@ -231,7 +231,7 @@ impl PhysicalPlanner {
     ) -> Result<Arc<dyn PhysicalExpr>, ExecutionError> {
         match spark_expr.expr_struct.as_ref().unwrap() {
             ExprStruct::Add(expr) => {
-                // TODO respect eval mode
+                // TODO respect ANSI eval mode
                 // https://github.com/apache/datafusion-comet/issues/2021
                 // https://github.com/apache/datafusion-comet/issues/536
                 let _eval_mode = from_protobuf_eval_mode(expr.eval_mode)?;
```
Comment (andygrove) [inline native/core/src/execution/planner.rs]: Please remove the leading `_` from the variable name now that we are using the variable

File: native/spark-expr/src/math_funcs/checked_arithmetic.rs [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+    for i in 0..len {
+        if left.is_null(i) || right.is_null(i) {
+            builder.append_null();
+        } else {
+            let l = left.value(i);
+            let r = right.value(i);
+
+            match op {
+                "checked_add" => builder.append_option(l.add_checked(r).ok()),
```
Comment (andygrove) [inline native/spark-expr/src/math_funcs/checked_arithmetic.rs]: Performing this `match` operation on every row will be expensive. It would be better to invert this and to the `match` once and then have a different `for` loop for each operation, something like this:

```rust
match op {
  "checked_add" =>
    for i in 0..len {
      ...
    }
  "checked_sub" =>
    for i in 0..len {
      ...
    }

**Comment by andygrove** [comment]
@coderfender I took a first pass through this, and I think this is looking good :+1:

File: native/core/src/execution/planner.rs [hunk 3]
```diff
@@ -878,6 +879,7 @@ impl PhysicalPlanner {
         return_type: Option<&spark_expression::DataType>,
         op: DataFusionOperator,
         input_schema: SchemaRef,
+        fail_on_overflow: bool,
```
Comment (andygrove) [inline native/core/src/execution/planner.rs]: Maybe consider passing in the eval mode here instead of a boolean? We'll eventually need to support all three modes.

Comment (coderfender) [reply native/core/src/execution/planner.rs]: Sure thats a great idea!

**Comment by coderfender** [comment]
One of the TPC-H check failed with a network exception. @andygrove could you please re trigger that workflow whenever you get a chance? 
Thank you

**Comment by coderfender** [comment]
@andygrove  , thank you for restarting the failed job and glad to see that the checks have all passed. Please review once you get a chance and let me know if you think we need further changes. 

Thank you

File: native/spark-expr/src/math_funcs/checked_arithmetic.rs [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        "checked_div" => {
+            for i in 0..len {
+                if left.is_null(i) || right.is_null(i) {
+                    builder.append_null();
+                } else {
+                    builder.append_option(left.value(i).div_checked(right.value(i)).ok());
+                }
+            }
+        }
+        _ => todo!("Unsupported operation: {}", op),
```
Comment (andygrove) [inline native/spark-expr/src/math_funcs/checked_arithmetic.rs]: Could you return an `Err` here?

File: native/spark-expr/src/math_funcs/checked_arithmetic.rs [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+    let left = &args[0];
+    let right = &args[1];
+
+    let (left_arr, right_arr): (ArrayRef, ArrayRef) = match (left, right) {
+        (ColumnarValue::Array(l), ColumnarValue::Array(r)) => (Arc::clone(l), Arc::clone(r)),
+        (ColumnarValue::Scalar(l), ColumnarValue::Array(r)) => {
+            (l.to_array_of_size(r.len())?, Arc::clone(r))
+        }
+        (ColumnarValue::Array(l), ColumnarValue::Scalar(r)) => {
+            (Arc::clone(l), r.to_array_of_size(l.len())?)
```
Comment (andygrove) [inline native/spark-expr/src/math_funcs/checked_arithmetic.rs:126]: We may eventually want to have a specialized version of the kernel for the scalar case to avoid the overhead of creating an array from the scalar. This does not need to happen as part of this PR, though.

Comment (coderfender) [reply native/spark-expr/src/math_funcs/checked_arithmetic.rs:126]: Sure ! I will create a follow up enhancement to track changes for a scalar impl. Thank you for the feed back @andygrove .

File: native/spark-expr/src/math_funcs/checked_arithmetic.rs (hunk 4 above)
Comment (coderfender) [reply native/spark-expr/src/math_funcs/checked_arithmetic.rs]: Thats valuable feedback . I updated code to return `Err` instead of `todo`

**Review by andygrove — APPROVED** [review]
LGTM. Thanks @coderfender

**Comment by andygrove** [comment]
@coderfender could you fix the conflicts?

**Comment by coderfender** [comment]
Thank you very much for the approval @andygrove . I am going to push a fix shortly after fixing conflicts.

**Comment by coderfender** [comment]
@andygrove  , There is a test failure with the below error after rebase with main branch. I am currently investigating the failure and patch a potential fix 

```Job aborted due to stage failure: Task 0 in stage 3.0 failed 1 times, most recent failure: Lost task 0.0 in stage 3.0 (TID 3) (172.16.208.49 executor driver): org.apache.comet.CometNativeException: Fail to deserialize to native operator with reason: failed to decode Protobuf message: Literal.is_null: Expr.expr_struct: Projection.project_list: Operator.op_struct: invalid wire type: LengthDelimited (expected Varint).
```

**Comment by coderfender** [comment]
@andygrove the checks have all passed. Thank you for your approval .Please merge once you get a chance

**Comment by coderfender** [comment]
Thank you very much for merging feature branch Andy . I created a new issue to extend on these changes and support ANSI mode for above arithmetic operations : #2137   (and raised a WIP PR #2136 )
